### PR TITLE
feat(mcp): support concurrent multi-project agent workflows

### LINKED ISSUE #127: fix: no headless way to set the MCP target with multiple projects connected; all MCP calls blocked (verified on 0.4.0)
## What happened?

We run AI coding agents in parallel, each in its own git worktree of the same project, each opening its own `--headless` editor. On 0.4.0, headless support is impressively close to complete: script diagnostics work (including on editors launched with non-default `--lsp-port`), `write_or_update_file` reports accurate post-write diagnostics, `run_scene_edit_script` edits scenes, and `runtime_session` even boots real runtime sessions with no GUI anywhere. That makes Fennara genuinely attractive for automation and CI.

The one remaining blocker is target selection. As soon as a second editor connects to the daemon, the active project becomes `none` and every MCP call from every client fails with:

```
Multiple Fennara projects are open. In the Fennara dock, choose Set as MCP target for the project you want to control.
```

The dock is the only resolution path (the FAQ confirms this), and it is a rendered GUI panel, so a `--headless` editor can never show it. We verified on 0.4.0 that no MCP tool exposes a project/session targeting parameter (`tools/list` schemas checked; the `session_id` on `runtime_session`/`runtime_script` refers to runtime sessions, not project routing). The result is a machine-wide deadlock of all Fennara MCP tooling with no programmatic recovery; the only way out is quitting editors until exactly one remains. We hit this in practice: one extra connected editor blocked every other client's calls for the ~13 minutes it stayed open.

## Expected behavior

A headless-compatible way to resolve the target. Any of these would unblock automation, roughly in order of usefulness:

1. An optional `project_path` (or session id) parameter on MCP tool calls, routed by the daemon. This is the most general fix: parallel clients could each address their own project without any global state at all.
2. Exposing the existing internal target-selection message (`set_mcp_target` in the daemon's godot bridge) through a programmatic surface: an MCP tool, a daemon CLI command, or an env/config option.
3. A daemon option to auto-target, for example most-recently-connected project, as a fallback when the dock is unavailable.

Why this matters: multi-instance headless is exactly the setup that agent orchestration and CI need (one editor per worktree/branch). Everything else already works headless in 0.4.0; this single UI-only step is what currently rules out parallel use. Option 1 would also remove a per-machine global bottleneck even for GUI users running two projects side by side.

## Steps to reproduce

1. Launch editor A: `godot --editor --headless --path /path/to/project-a --lsp-port 6006` and wait for the daemon to connect it. MCP calls work and target project A.
2. Launch editor B on a second project (a git worktree of the same game, or any minimal project with the addon, reproduces this): `godot --editor --headless --path /path/to/project-b --lsp-port 6007`.
3. Once B connects, call any MCP tool (read-only ones included, e.g. `project_settings`): every call now fails with the "Multiple Fennara projects are open" error, and `fennara_status` reports `Active project: none` with both projects listed under connected projects.
4. No headless mechanism exists to pick a target; quitting one editor restores auto-selection.

## Environment

- OS: macOS 26.5
- Godot version: 4.6.2-stable (official)
- Fennara version: 0.4.0 (daemon, MCP server, and project addon all 0.4.0; reproduced today on this version)
- MCP app: Claude Code

## Logs or screenshots

```
Tool: project_settings
Status: failed
Error: Multiple Fennara projects are open. In the Fennara dock, choose Set as MCP target for the project you want to control.
```

`fennara_status` during the failure: `Active project: none`, `Connected projects: 2` (both projects listed and healthy, plugin 0.4.0 on both).

### PR DESCRIPTION
**Pull request by GenKerensky**
### What changed?

- Bind each MCP process to one Godot project through `--project-path`, `FENNARA_PROJECT_PATH`, or automatic discovery from its startup directory.
- Route bound calls through the shared daemon by canonical project identity without mutating the legacy dock-selected target, and fail closed when a binding is missing or ambiguous.
- Serialize daemon-managed playtests through one Runtime Slot with anonymous busy polling, owner-scoped operations, inactivity renewal, and an absolute lease.
- Document the architecture, setup, status, and multi-agent workflow across the canonical docs and all nine translated locales.

### Why?

Parallel agents currently share one daemon-global MCP target. With multiple Godot editors connected, headless agents either become blocked or risk depending on mutable global selection. Process-scoped binding lets each MCP connection remain pinned to its own repository or worktree while preserving the existing unbound behavior.

Managed game runs remain a machine-wide scarce resource for now. The Runtime Slot lets agents poll and take turns without turning ordinary contention into an MCP error; bounded validation and standalone screenshots remain concurrent.

Closes #127

> [!WARNING]
> Platform-specific behavior has not yet been tested on Windows or macOS. Native build, runtime, and live-routing verification was performed on Linux only.
>
> One Fennara MCP connection may be shared only by agents intentionally working in the same Project Root. If an MCP host moves a subagent into an isolated worktree while reusing the parent's connection, Fennara remains bound to the parent's Project Root because the tool request contains no trustworthy caller-workspace identity. That isolated subagent must receive its own project-bound Fennara MCP connection.

### How did you verify it?

- `cd local && cargo test --workspace --locked`
- `cd local && cargo build --release --workspace --locked`
- `cd local && cargo fmt --all -- --check`
- `rust-analyzer diagnostics local`
- `node scripts/sync-doc-navigation.mjs --check`
- `node scripts/check-doc-i18n.mjs`
- Linux GDExtension build/link and Godot headless editor smoke test
- Live Linux smoke test with two independently bound Godot projects through one daemon
- `git diff --check origin/main...HEAD`

### Checklist

- [x] I kept the change focused
- [x] I updated docs if behavior changed
- [x] I did not include unrelated changes
- [x] I explained any workflow or release-process changes (none)

## Summary by CodeRabbit

* **New Features**
  * MCP connections can bind to specific Godot projects for isolated multi-project and worktree workflows.
  * Added project-aware routing status, binding discovery, reconnection behavior, and clearer diagnostics.
  * Added shared Runtime Slot management with ownership, busy-state reporting, renewable leases, expiry handling, and cleanup.
  * Scene validation and standalone screenshots can run without occupying the interactive Runtime Slot.
  * Added runtime limits, startup validation, and owner-only session controls.

* **Documentation**
  * Added setup, architecture, troubleshooting, FAQ, and multi-agent worktree guidance.
  * Expanded guidance for concurrent workflows, project binding, runtime polling, and lease behavior.

### CODE REVIEW

**Comment by GITushar23** [comment]
Hey, really nicely documented pr. I just have one confusion with this approach.
You expect user who want to use multi agent workflow to add a mcp server configuration in all of there worktrees or project.
But this work if say your claude code session decides to spin up new claude code session  in the respective dir, and the central one controls all session in this case this implementation works fine. 
But say a Opus decides to use its in build Subagent tool that subagent will inherit the mcp process from its parent so it will edit same project, now opus may intend it to edit the same project or the diff project this will not be handled??

This is my confusion i am not that much into multi agent workflows but if u can clarify how things are supposed to work or creating new worktrees and spawning seperate claude session is how it works then its fine.

**Comment by GenKerensky** [comment]
That might be a legitimate concern. I've only smoke tested this change with a single agent running in 3 different Godot projects simultaneously. All sub agents do inherit their parent agent's MCP config as far as I know so that would cause routing collisions if sub agents were spun up in their own worktrees.

This is why I have the PR in draft mode right now, wanted to get it out there, but I'm not sure I've covered every corner case. I'll test with Claude and Codex soon and see if I need to re-think the design.

I mostly manage my own worktrees, intentionally spinning them up for feature branches rather than as a part of an automated, sub-agent, orchestration process.

**Comment by GenKerensky** [comment]
Checking the current docs for several harnesses, and it looks like this only a problem in Claude when using very specific, opt-in workflows, and it is more of a risk in Grok build where the orchestrator can choose to worktree isolate a subagent on its own:

---
> AI summarized

### Claude Code

Ordinary Claude subagents start in the parent conversation’s working directory. Worktree isolation is opt-in through `isolation: worktree`, asking Claude to use worktrees, or built-in workflows such as `/batch`.

Claude also explicitly documents that named MCP server references reuse the parent session’s connection. Its worktree enforcement protects native file-edit, shell, and Git operations, but does not mention MCP calls.

Therefore, a worktree-isolated Claude subagent using an inherited Fennara server is affected: its native tools operate in the child worktree, while Fennara remains bound to the parent project.

Sources:

- [Claude subagent working directories and isolation](https://code.claude.com/docs/en/sub-agents)
- [Isolating Claude subagents with worktrees](https://code.claude.com/docs/en/worktrees#isolate-subagents-with-worktrees)
- [Claude MCP connection sharing](https://code.claude.com/docs/en/sub-agents#scope-mcp-servers-to-a-subagent)
- [What Claude’s worktree isolation actually enforces](https://code.claude.com/docs/en/worktrees#how-claude-code-enforces-isolation)
- [Claude parallel-agent modes, including `/batch`](https://code.claude.com/docs/en/agents)

### Codex

Codex documents subagents as threads within the current chat. Its worktrees are chat-level environments: users start or hand off a chat into a worktree, and managed worktrees are normally dedicated to one chat.

I found no documented Codex option that automatically places an individual subagent into a different worktree. Separate worktree chats are the intended isolation boundary, so this specific inherited-connection problem is not documented for Codex’s built-in subagents.

Sources:

- [Codex subagents](https://learn.chatgpt.com/docs/agent-configuration/subagents)
- [Codex Git worktrees](https://learn.chatgpt.com/docs/environments/git-worktrees)

### Grok Build

Grok Build has essentially the same potential edge case as Claude. Its model-facing `spawn_subagent` tool supports `isolation: worktree`; shared-workspace isolation is the default, but the parent model or an agent configuration can select a worktree.

Grok also documents that subagents inherit the parent’s already-connected MCP servers by default. A worktree-isolated Grok subagent can therefore operate in a child checkout while Fennara remains bound to its parent’s project.

Sources:

- [Grok Build worktrees](https://docs.x.ai/build/features/worktrees)
- [Grok Build’s detailed subagent and MCP-inheritance contract](https://github.com/xai-org/grok-build/blob/main/crates/codegen/xai-grok-pager/docs/user-guide/16-subagents.md)

### OpenCode

OpenCode documents automatically invoked subagents as child sessions in the current project. Its built-in agent schema does not document a worktree-isolation option. Worktree automation is listed separately through ecosystem plugins rather than as default subagent behavior.

This means the issue is not documented for stock OpenCode, although a custom worktree orchestration plugin could recreate it.

Sources:

- [OpenCode agents](https://opencode.ai/docs/agents/)
- [OpenCode ecosystem and worktree plugin](https://opencode.ai/docs/ecosystem/)

---

I think this should be documented explicitly as a host-integration boundary. Deterministic, connection-scoped binding is more reliable than an architecture where you expect sub agents to accurately report their workspace directory to the parent's mcp. Claude Code and Grok Build users should know that worktree-isolated subagents must receive their own Fennara MCP connection rather than inheriting the parent one. Might be worth adding it to the CLAUDE/AGENTS.md fennara block to never use worktree isolation in subagent runs.

That said, subagent workflows are becoming popular and if harnesses aren't going to isolate MCPs in worktrees, we might be able to come up with a different architecture that supports that reliably. If you think it's worth it.

**Comment by GITushar23** [comment]
@coderabbitai full review

**Comment by GenKerensky** [comment]
Rebased onto main and resolved conflicts with #150 (`user_args` for runtime sessions) while keeping this PR's multi-project Runtime Slot / binding behavior.

**Comment by GenKerensky** [comment]
Fixed Local Build failures on macOS/Windows: status assertions now match markdown-escaped paths, non-UTF-8 filename tests are Linux-only, and the socketpair fixture isolates HOME so macOS does not read a real daemon token.

**Comment by GenKerensky** [comment]
Documented the inherited-MCP issue for worktree-isolated subagents under [Worktree-Isolated Subagents](https://github.com/fennaraOfficial/fennara-godot-ai/blob/feature/multi-project-support/docs/multi-agent-worktrees.md#worktree-isolated-subagents). Claude Code `isolation: worktree` and Grok Build `spawn_subagent` worktree isolation keep Fennara bound to the parent project, so the child needs its own Fennara MCP connection or must stay in the parent tree.

File: fennara-cpp/src/executor/runtime_session_batch.cpp [hunk 1]
```diff
@@ -133,173 +144,111 @@ void FennaraExecutor::_on_runtime_session_complete(uint64_t batch_generation) {
...
     if (!done) {
-        godot::SceneTree *tree = get_tree();
-        if (tree) {
-            godot::Ref<godot::SceneTreeTimer> timer =
-                tree->create_timer(kPollSeconds);
-            timer->connect("timeout", callable_mp(this, &FennaraExecutor::_on_runtime_session_complete).bind(batch_generation));
-        }
+        _schedule_runtime_session_poll(batch_generation);
         return;
     }
```
[the remark this answers is not in the document]
Comment (GenKerensky) [reply fennara-cpp/src/executor/runtime_session_batch.cpp:157]: Addressed in 2a2ba071321cd9cdba48caef9057e3b15e7826d6: if SceneTree poll scheduling fails, join the worker and complete the phase from its result instead of dropping the only remaining poll.

File: fennara-cpp/src/tools/runtime_session/runtime_session.cpp [hunk 2]
```diff
@@ -237,16 +277,17 @@ godot::Dictionary FennaraRuntimeSessionTool::execute(
...
-        godot::Dictionary result = daemon_status(session_id);
-        result["tool_name"] = "runtime_session";
-        if (result.has("raw_log_path") && !result.has("log_path")) {
-            result["log_path"] = result["raw_log_path"];
+        godot::Dictionary result = daemon_status(root.path(), session_id);
+        attach_runtime_result_fields(result);
+        remember_owned_session(result);
+        if (!(bool)result.get("slot_acquired", false)) {
+            clear_active_daemon_session_id_if_matches(session_id);
         }
```
[the remark this answers is not in the document]
Comment (GenKerensky) [reply fennara-cpp/src/tools/runtime_session/runtime_session.cpp:319]: Addressed in 2a2ba071321cd9cdba48caef9057e3b15e7826d6: remembered owner session id is now cleared only after a successful daemon status reply with slot_acquired=false, matching query_slot_status.

File: local/README.md [hunk 3]
```diff
@@ -18,9 +18,20 @@ Endpoints:
 
 - `GET /health`: daemon health.
 - `GET /status`: daemon status plus connected Godot plugin metadata.
+- privileged bound status: resolves one MCP process's canonical Project Root
+  against connected Godot editor sessions.
```
[the remark this answers is not in the document]
Comment (GenKerensky) [reply local/README.md]: Addressed in 2a2ba071321cd9cdba48caef9057e3b15e7826d6: named the bound status endpoint as `POST /status/bound` in local/README.md and the i18n copies.

**Comment by GenKerensky** [comment]
Skipped wrapping `matching_bound_session_ids` / `ProjectRoot::resolve_absolute` in `spawn_blocking`. Bound routing resolves local Godot project directories on local disk; a `spawn_blocking` hop on every bound tool call adds scheduler overhead for a path that is already cheap. If network mounts become a real issue, cache canonical identity on editor hello rather than offloading each lookup.

**Comment by GITushar23** [comment]
hey @GenKerensky  really really sorry i took so long for review this is my placement season and i am doing so many other things its becoming little hard to manage so sorry to keep you waiting.
few bugs I have faced I will try and make some commits 
max_run_seconds: 90 becomes 90.0 inside Godot. The daemon expects Option<u64>, rejects the request, and Fennara reports Status: unknown. This is a PR regression.
Not in this PR but is a bug for now that needs fixing just telling here , Fennara kills the game after 5 seconds, but the real Godot Mono process needed about 8 seconds to become ready.

**Comment by GenKerensky** [comment]
@GITushar23 no sweat, that's just the nature of open source, everyone's a volunteer. Guess I should also validate this in Godot Mono, only used the vanilla version.

**Comment by GenKerensky** [comment]
@GITushar23 the `max_run_seconds: 90` → `90.0` regression is fixed in 0a19978. Godot JSON-stringifies whole numbers as floats; the daemon now accepts whole JSON numbers, and the C++ client sends an integer. Fractional values like `90.5` still fail closed.

The 5-second ready wait you hit on Godot .NET is `STARTUP_READY_TIMEOUT_MS` and is separate from this PR's lease. I left it alone as you noted; worth a follow-up if .NET C# projects regularly need longer than 5s to print the ready marker.

**Comment by GenKerensky** [comment]
@GITushar23 follow-up on the Godot .NET ready wait: `16953bf` raises the startup deadline from 5s to 20s. If `FENNARA_RUNTIME_SESSION_READY` / `FENNARA_RUNTIME_ORIENTATION_NOTE` still miss that window, the daemon still kills and reaps (`startup_timeout`).

Verified on this machine with Arch `godot` 4.7.2 and `godot-mono` 4.7.2: both reported ready, kept `max_run_seconds: 90` (including JSON `90.0` over HTTP), and stopped cleanly.

File: fennara-cpp/src/tools/runtime_session/runtime_session.cpp [hunk 4]
```diff
@@ -212,18 +252,36 @@ godot::Dictionary FennaraRuntimeSessionTool::execute_start_after_preflight(
...
+        int64_t seconds = 0;
+        bool whole = false;
+        if (raw.get_type() == godot::Variant::INT) {
+            seconds = int64_t(raw);
+            whole = seconds >= 0;
+        } else if (raw.get_type() == godot::Variant::FLOAT) {
+            double value = double(raw);
+            whole = std::isfinite(value) && value >= 0.0 && value == std::floor(value) &&
+                    value <= static_cast<double>(INT64_MAX);
+            seconds = static_cast<int64_t>(value);
```
[the remark this answers is not in the document]
Comment (GenKerensky) [reply fennara-cpp/src/tools/runtime_session/runtime_session.cpp]: Addressed in f9eef0fa5f313ac4b99515a77409e5c739f92ba2: FLOAT max_run_seconds is converted to int64_t only after it is a finite whole number strictly below 2^63, so NaN, infinity, and the rounded INT64_MAX double no longer hit undefined behavior.

**Comment by GITushar23** [comment]
The daemon now waits up to 20 seconds for Godot readiness, plus up to 3 seconds for startup capture. The C++ client still stops waiting after 10 seconds in runtime_session.cpp

**Comment by GenKerensky** [comment]
@GITushar23 good catch. `start` now uses a 25s HTTP wait in the C++ client (20s ready + 3s capture + margin). Status and stop stay at 10s. In 1ab76aa.
